src/app/launcher/shared/project-name.validator.ts

```diff
@@ -10,7 +10,9 @@ import { DependencyCheckService } from '../service/dependency-check.service';
     forwardRef(() => ProjectNameValidatorDirective), multi: true }]
 })
 export class ProjectNameValidatorDirective implements Validator {
-  private pattern = /^[a-z][a-z0-9-]{3,63}$/;
+  // allows only '-', '_' and 4-40 characters (must start with alphabetic and end with alphanumeric)
+  // no continuous '-' or '_' is allowed
+  private pattern = /^[a-zA-Z](?!.*--)(?!.*__)[a-zA-Z0-9-_]{2,38}[a-zA-Z0-9]$/;
```

Comment:
could we have a few unit tests for this pattern?

Reply:
@ia3andy there is no spec file itself for this directive, for pattern I can add but it'll be tough with directive.

Reply:
@invincibleJai Have a look at `properties.spec.ts`, you could just make this pattern public static and readonly so you can test it without actually createing the directive?

Reply:
@ia3andy pls take a look!!